python/fusion_engine_client/messages/configuration.py

```diff
@@ -21,6 +21,199 @@ class ConfigType(IntEnum):
+    UP = 4,
+    ## Aligned with vehicle -z axis.
+    DOWN = 5,
+    ## Error value.
+    INVALID = 255
+
+
+class _ConfigClassGenerator:
+    """!
+    @brief Internal class for generating ConfigClass children.
+
+    These classes consist of 3 pieces:
+    - The ConfigType associated with the class.
+    - An accessor class. This is the class that's used to get/set the values for the config object.
+    - A serialization class. This is the class that (de)serializes the data for the config object.
+
+    To generate a ConfigClass:
+    - Declared as a child of NamedTuple. The NamedTuple defines the fields.
+    - Add a create_config_class decorator that takes the ConfigType and serialization Construct associated with the class.
+
+    For example:
+    ```{.py}
+        gen = _ConfigClassGenerator()
+
+        class Foo(NamedTuple):
+            x: int
+
+        FooConstruct = Struct(
+            "x" / Int32ul
+        )
+
+        @gen.create_config_class(ConfigType.BAR, FooConstruct)
+        class Bar(Foo)
+    ```
+    Would create a new ConfigClass Bar. Messages with ConfigType.BAR will attempt to (de)serialize to Bar. This
+    serialization is defined by FooConstruct. The user accessible fields are defined in Foo.
+    """
+    class ConfigClass:
+        """!
+        @brief Abstract base class for accessing configuration types.
+        """
+        @classmethod
+        def GetType(cls) -> ConfigType:
+            raise ValueError('Accessing `GetType()` of base class')
+
+    def __init__(self):
+        # Gets populated with the mappings from ConfigType to constructs.
+        self.CONFIG_MAP = {}
+
+    def create_config_class(self, config_type, construct_class):
+        """!
+        @brief Decorator for generating ConfigClass children.
+
+        @copydoc _ConfigClassGenerator
+        """
+        def inner(config_class):
+            # Make the decorated class a child of ConfigClass. Add the GetType method.
+            class InnerClass(config_class, self.ConfigClass):
+                @classmethod
+                def GetType(cls) -> ConfigType:
+                    return config_type
+            InnerClass.__name__ = config_class.__name__
+
+            # Register the construct with the MessageType.
+            self.CONFIG_MAP[config_type] = NamedTupleAdapter(InnerClass, construct_class)
+
+            return InnerClass
+        return inner
+
+    class Point3F(NamedTuple):
+        """!
+        @brief 3D coordinate specifier, stored as 32-bit float values.
+        """
+        x: float = 0
+        y: float = 0
+        z: float = 0
+
+    # Construct to serialize Point3F.
+    Point3FConstruct = Struct(
+        "x" / Float32l,
+        "y" / Float32l,
+        "z" / Float32l,
+    )
+
+    class IntegerVal(NamedTuple):
+        """!
+        @brief Integer value specifier.
+        """
+        value: int
+
+    # Construct to serialize 32 bit IntegerVal types.
+    UInt32Construct = Struct(
```

Comment:
This says uint but the line below says int. Which is it?

Reply:
The `u` in Int32ul is for unsigned.

Reply:
[int][32bit][unsigned][little endian]